### PR TITLE
Add log utils for Auto plugin

### PR DESCRIPTION
**Pull request by tiger100256-hu**
### Details:
 - *Add log utils for plugin
 - *Add log for select device*
 - *Add log for device config*
 - *Add test case*
 

### Tickets:
 - *68150*

### CODE REVIEW

**Comment by myshevts** [comment]
finally if we go out with the logging utility (with levels), it should be general (reside in common) and thus re-usable by other plugins
adding more plugin people: @dmitry-gorokhov , @vladimir-paramuzov , @ilya-lavrenov , @vinograd47

**Comment by vinograd47** [comment]
I'm not sure, which input do you expect. Right now each plugin implements the loggings on its own. Even if some common mechanism is introduced in the core library, I'm not sure if the plugins will switch to it (since they already have something and it works). Or this switch might be just a low priority task.

As for current implementation - I have one concern. It looks like the logger is a single global singleton. While the OpenVINO API allows to have granular log level control (eg. different levels for various `LoadNetwork` call).

**Comment by peterchen-intel** [comment]
WW45.1 review 
1. More details in description
2. log functions' name and file name
3. All log related should be in log implementation source code file

File: inference-engine/src/multi_device/utils/singleton.cpp
Comment (tiger100256-hu) [reply inference-engine/src/multi_device/utils/singleton.cpp]: remove this redundant code, help to have a look again

File: inference-engine/src/multi_device/utils/singleton.hpp
Comment (tiger100256-hu) [reply inference-engine/src/multi_device/utils/singleton.hpp]: remove this redundant code, help to have a look again

File: inference-engine/src/multi_device/multi_device_exec_network.cpp
Comment (myshevts) [inline inference-engine/src/multi_device/multi_device_exec_network.cpp:296]: how this relates to log introduced in this PR?

File: inference-engine/src/multi_device/utils/log.hpp
Comment (myshevts) [inline inference-engine/src/multi_device/utils/log.hpp:60]: this is in fact really good idea, to be able to enable the log thru the env, without config change

**Comment by myshevts** [comment]
@vinograd47  and @ilya-lavrenov , I still believe we can generalize the code and let any plugin use that (with the common config key and/or env).

**Review by myshevts — COMMENTED** [review]
let's close on the generalization, I think overall the functionality is quite useful to become common for IE..
Also e.g. user may want to include some LOG_* in his code to see correlation with logs from OV, so maybe the code (macros) should be made public API

**Comment by vinograd47** [comment]
In that case you need more features. 

For example, this implementation is based on `std::snprintf`, so it supports only standard C-like formatting. Logger class in our plugin has rich and extensible formatting support similar to `std::format` from C++20.

Another example - colorized output. Looks like this implementation supports only ANSI-like terminals. Our plugin supports both ANSI and Windows terminals.

Summarizing, if we want to have single Logger implementation and force plugins to use them, we need first to align on features.

**Comment by myshevts** [comment]
> In that case you need more features.
> Summarizing, if we want to have single Logger implementation and force plugins to use them, we need first to align on features.

@vinograd47, why wouldn't fully productize/generalize VPU's logger then (and move to common)?
We need the logging by the 22.1, so if this not an option, we will continue with out impl and let's consider refactoring of any impl for 22.2 then. In tis case would you pls have a general look on our impl?

**Comment by vinograd47** [comment]
> why wouldn't fully productize/generalize VPU's logger then (and move to common)?

The Logger I was talking about is implemented in private plugin. Also it use some staff from 3rd-party library, which is used in the plugin.

I don't mind if you add some logging utilities to core part, but it shouldn't be a hard requirement for all plugins to switch to it (at least in nearest future).

**Comment by myshevts** [comment]
> but it shouldn't be a hard requirement for all plugins to switch to it (at least in nearest future).
Sure, I suggest until we have common/aligned/approved-by-arch-review logger, some plugins can have its own impl.

@wangleis, so looks like in short-term the only option is to keep the current impl as AUTO-specific, unless your team will take the generalization/review with all CPU/GPU plugins owners (archForum?) ...what would be the possible timeline then (likely not for 22.1/ in whihc the AUTO needs some solution).

**Comment by wangleis** [comment]
@myshevts agree with you to implement logging in AUTO device for 2022.1 release first.

File: inference-engine/src/multi_device/multi_device_exec_network.cpp
Comment (myshevts) [reply inference-engine/src/multi_device/multi_device_exec_network.cpp:296]: pls close on this item, this seems to be a logic change unrelated to the logging

Comment (tiger100256-hu) [reply inference-engine/src/multi_device/multi_device_exec_network.cpp:296]: yes, I know it's confused to add this code. I want to print the information of GPU here, when switch from CPU to GPU.
I also thought if could add the logs in the loading thread, but GPU may load failed [8006](https://github.com/openvinotoolkit/openvino/pull/8006), so add logs in "WaitActualNetworkReady" maybe the best way.

File: inference-engine/src/multi_device/utils/log.hpp
Comment (tiger100256-hu) [reply inference-engine/src/multi_device/utils/log.hpp]: enable it with env variable "AUTO_LOG_LEVEL"

File: inference-engine/src/multi_device/multi_device_exec_network.cpp
Comment (tiger100256-hu) [reply inference-engine/src/multi_device/multi_device_exec_network.cpp:298]: add comments in code

File: inference-engine/src/multi_device/CMakeLists.txt
Comment (ilya-lavrenov) [inline inference-engine/src/multi_device/CMakeLists.txt:8]: why not just `GLOB_RECURSE`?

File: inference-engine/src/multi_device/multi_device_exec_network.cpp
Comment (myshevts) [reply inference-engine/src/multi_device/multi_device_exec_network.cpp:298]: why it cannot be print when the network loading is completed? (in the async task)

Comment (wangleis) [reply inference-engine/src/multi_device/multi_device_exec_network.cpp:298]: We would like to print full logic path include QueryNetwork success and LoadNetwork fail.

File: inference-engine/src/multi_device/utils/log.hpp
Comment (tiger100256-hu) [reply inference-engine/src/multi_device/utils/log.hpp]: change AUTO_LOG_LEVEL to OPENVINO_LOG_LEVEL

File: inference-engine/src/multi_device/CMakeLists.txt
Comment (tiger100256-hu) [reply inference-engine/src/multi_device/CMakeLists.txt]: fix

File: inference-engine/src/multi_device/multi_device_exec_network.cpp
Comment (myshevts) [reply inference-engine/src/multi_device/multi_device_exec_network.cpp:377]: having a query/print code in the innermost scheduling loop is not acceptable.
from what I see it does always query the config, but just doesn't print that when the LOG_LEVEL is not high enough.

alos the configs are static, so not sure why we need to print them from the scheduling code

**Comment by tiger100256-hu** [comment]
please wait a moment , not merge it, thank you

File: inference-engine/src/multi_device/multi_device_exec_network.cpp
Comment (tiger100256-hu) [reply inference-engine/src/multi_device/multi_device_exec_network.cpp]: move the print into load task

File: inference-engine/src/multi_device/multi_device_plugin.cpp
Comment (peterchen-intel) [inline inference-engine/src/multi_device/multi_device_plugin.cpp:525]: Keep as original logic, create new PR if change is required.

File: inference-engine/src/multi_device/multi_device_plugin.cpp
Comment (peterchen-intel) [inline inference-engine/src/multi_device/multi_device_plugin.cpp:539]: Keep as original logic, create new PR if change is required.

File: inference-engine/src/multi_device/multi_device_plugin.cpp
Comment (peterchen-intel) [inline inference-engine/src/multi_device/multi_device_plugin.cpp:557]: Please check if this is necessary change for log enabling.

File: inference-engine/src/multi_device/multi_device_plugin.hpp
Comment (peterchen-intel) [inline inference-engine/src/multi_device/multi_device_plugin.hpp:59]: Please check if this is necessary change for log enabling.

**Review by peterchen-intel — CHANGES_REQUESTED** [review]

File: inference-engine/src/multi_device/multi_device_exec_network.cpp
Comment (myshevts) [inline inference-engine/src/multi_device/multi_device_exec_network.cpp:213]: why not _confMutex is locked

Comment (tiger100256-hu) [reply inference-engine/src/multi_device/multi_device_exec_network.cpp:215]: fix

**Review by myshevts — APPROVED** [review]

**Review by wangleis — APPROVED** [review]

**Review by peterchen-intel — APPROVED** [review]

**Pull request merged**
